### ISSUE TITLE
scale of the mine

### ISSUE THREAD

**Issue by Daerduo0610** [370025263/openmines#14]
Hi, thank you for your great work on this project!

I have a question regarding the scalability of your RL algorithm. In your experiments, what is the largest problem size you have tested? (e.g., in terms of loading sites number × unloading sites number × truck number)

I implemented a similar simulation environment and tested it on a different mine dataset. However, I found that the algorithm struggles to converge, or converges to solutions worse than a simple baseline. I'm wondering if this might be due to the problem scale I have tested being too large.

Would appreciate any insights or suggestions you might have!

**Comment by 370025263** [comment]
Hi @Daerduo0610 ～

Thanks for your interest and kind words about the project!

Regarding your question about scalability and problem size, in our experiments, we tested on a 5x5 site configuration (meaning 5 loading sites and 5 unloading sites) and varied the number of trucks from 1 to 120 (with different loading abilities).

<img width="548" alt="Image" src="https://github.com/user-attachments/assets/9fd69766-8ec3-4498-a7c8-e555329c3f27" />

It's indeed challenging to get RL algorithms to converge well and outperform strong baselines in complex dispatching environments. This was actually my very first RL work, and I struggled quite a bit with training the agent for a long time. Based on my experience, here are a few points that might be helpful:
- Use Existing Implementations: If you are new to RL, I highly recommend starting with existing, well-tested algorithm implementations (like those in libraries such as CleanRL, Stable Baselines3, etc.) rather than implementing everything from scratch. My initial self-implemented PPO actually performed worse than baselines.
- Utilize Traditional Algorithms as Guides: Traditional dispatching algorithms can be very strong baselines. Instead of just comparing against them, you might try utilizing their 'power' as a guide during RL training to reduce exploration costs. I tried this approach (using the output probabilities of a strong baseline policy as guidance/regularization in the loss function) with PPO, and it proved effective in stabilizing training and improving performance. For value-based methods like DQN, our colleague Xiaotong Zhang has explored a similar idea with an MBS approach, which you can find in the paper "Vehicle dynamic dispatching using curriculum-driven reinforcement learning".
- Focus on Implementation Details: Pay close attention to details like input normalization and reward normalization. These were crucial for stabilizing the training process for me; I initially suspected value function issues due to large reward variance before properly implementing these.


To give you a bit more context from my own journey: Initially, my own PPO implementation didn't converge well at all compared to baselines. Switching to a CleanRL implementation helped achieve convergence, but it was still difficult to surpass strong baselines. It was only after adding input and reward normalization, and incorporating the adaptive policy guidance (using probabilities from a strong baseline policy), along with using a dense reward structure, that the performance finally improved significantly. The algorithm became stable and consistently outperformed the baseline that guided it.

<img width="520" alt="Image" src="https://github.com/user-attachments/assets/66680806-0e1e-41e4-9aec-03f1ffa893ca" />

I also experimented with pre-training on data generated by a rule-based policy, but this approach didn't work well for me. My guess is it was due to significant 'data style' differences and not incorporating reward information during pre-training.

It took me several months to finally converge to a relatively satisfactory result, and even with these improvements, the final performance is still quite far from the AlphaGo-level dominance I initially hoped for in beating strong traditional methods.

You can check my RL paper here: https://arxiv.org/abs/2502.20845

Developing robust RL dispatchers is challenging but rewarding. I hope you can develop a powerful RL dispatcher!

Best regards,

Stone Meng.
